File: Leetcode July Challenge/C++/Lowest Common Ancestor of a Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL || root->val==p->val || root->val==q->val){
            return root;
        }
        
        TreeNode* a=NULL, *b=NULL;
        
        if(root->val > p->val && root->val > q->val){
            a=lowestCommonAncestor(root->left, p, q);
        }
        
        else if(root->val < p->val && root->val < q->val){
            b=lowestCommonAncestor(root->right, p, q);
        }
        else{
            a=lowestCommonAncestor(root->left, p, q);
            b=lowestCommonAncestor(root->right, p, q);
        }
        
        if(a==NULL && b==NULL){
            return NULL;
        }
        
        if(a!=NULL && b==NULL){
            return a;
        }
        
        if(a==NULL && b!=NULL){
            return b;
        }
        
        if(a!=NULL && b!=NULL){
            return root;
        }
        return root;
        
    }
};

File: Leetcode July Challenge/Python/Lowest Common Ancestor of a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
        
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root == None or root == p or root == q:
            return root
        
        l, r = None, None
        if root.val > p.val and root.val > q.val:
            l = self.lowestCommonAncestor(root.left, p, q)
        
        elif  root.val < p.val and root.val < q.val:
            r = self.lowestCommonAncestor(root.right, p, q)
        
        else:
            l = self.lowestCommonAncestor(root.left, p, q)
            r = self.lowestCommonAncestor(root.right, p, q)
        
        if l and r:
            return root
        
        return l if l else r
